### PR TITLE
Automatically set `ParameterType::BOOLEAN` for booleans

### PR DESCRIPTION
**Pull request by fritzmg**
Contao 4 already has support for `boolean` SQL fields. However, there is still one issue remaining: if you try to copy records, you might encounter the following error:

```
Doctrine\DBAL\Exception\DriverException:
An exception occurred while executing a query: SQLSTATE[22007]: Invalid datetime format: 1366 Incorrect integer value: '' for column `tl_content`.`foobar` at row 1
```

This can be reproduced by the following DCA:

```php
// contao/dca/tl_content.php
$GLOBALS['TL_DCA']['tl_content']['fields']['foobar'] = [
    'eval' => ['doNotCopy' => true],
    'sql' => ['type' => 'boolean', 'default' => false],
];
```

Update the database and then try to copy a page article that contains content elements.

This happens because the PDO driver automatically casts everything to strings, if no type is given - which for `false` means an empty string - which in turn is not compatible with a `boolean` (`tinyint(1)`) field.

To fix this we need to backport [this](https://github.com/contao/contao/blob/110d32fd52f5abfb58a07d15813d09dda7292e8e/core-bundle/contao/library/Contao/Database/Statement.php#L243-L248) from https://github.com/contao/contao/pull/4797.

### CODE REVIEW

**Comment by ausi** [comment]
But is this compatible the other way round?
Meaning does querying for `false` still work with `CHAR(1)` fields that are set to the empty string?

**Comment by fritzmg** [comment]
You mean something like

```php
Database::getInstance()
    ->prepare("SELECT * FROM tl_content WHERE invisible = ?")
    ->execute([false])->fetchAllAssoc()
;
```

? Yeah, that won't work anymore :/

**Comment by fritzmg** [comment]
But how likely is this? Everywhere in the core we don't use a prepared statement for queries like that (I think).

But I guess we should not break this in 4.13 and instead extensions need to apply a workaround like this: https://github.com/terminal42/contao-notification_center/pull/325 or this:

```php
'sql' => ['type' => 'boolean', 'default' => 0],
```

**Comment by ausi** [comment]
> Yeah, that won't work anymore :/

I think we could fix that be using the correct type (integer) in `$arrTypes` as `WHERE invisible = 0` should work for both cases but because the default type is string it would currently query `WHERE invisible = '0'`

**Comment by fritzmg** [comment]
Yes, that works 👍 (tested both situations)

File: core-bundle/src/Resources/contao/library/Contao/Database/Statement.php [hunk 1]
```diff
@@ -257,17 +258,26 @@ public function query($strQuery='', array $arrParams = array(), array $arrTypes
...
 				if (\is_string($varParam) || \is_bool($varParam) || \is_float($varParam) || \is_int($varParam) || $varParam === null)
 				{
 					return $varParam;
 				}
 
 				return serialize($varParam);
 			},
-			$arrParams
+			array_keys($arrParams),
+			array_values($arrParams)
```
Comment (ausi) [inline core-bundle/src/Resources/contao/library/Contao/Database/Statement.php]: If it is possible that `$arrParams` is not a `list` we have to use array_values() for the first parameter as well.

```suggestion
			array_keys(array_values($arrParams)),
			array_values($arrParams)
```

If not, array_values would not be needed at all:

```suggestion
			array_keys($arrParams),
			$arrParams
```

File: core-bundle/src/Resources/contao/library/Contao/Database/Statement.php [hunk 2]
```diff
@@ -257,17 +258,26 @@ public function query($strQuery='', array $arrParams = array(), array $arrTypes
 			throw new \Exception('Empty query string');
 		}
 
+		$arrTypes += array_fill(0, \count($arrParams), null);
```
Comment (ausi) [inline core-bundle/src/Resources/contao/library/Contao/Database/Statement.php]: Are we sure using `null` here is well supported?
Shouldn’t we use `ParameterType::STRING` instead?

Comment (fritzmg) [reply core-bundle/src/Resources/contao/library/Contao/Database/Statement.php]: We should, yes. There is actually a deprecation in place:

```
Using NULL as prepared statement parameter type is deprecated. Omit or use ParameterType::STRING instead.
```

File: core-bundle/src/Resources/contao/library/Contao/Database/Statement.php (hunk 1 above)
Comment (fritzmg) [reply core-bundle/src/Resources/contao/library/Contao/Database/Statement.php]: Good point. Actually I think we would need to support both cases, don't we? i.e.

```php
->prepare("SELECT * FROM tl_content WHERE invisible = ?")
->execute([false])
```
```
->prepare("SELECT * FROM tl_content WHERE invisible = :invisible")
->execute(['invisbile' => false])
```

So your second suggestion is correct. I'll do some further adjustments.

**Review by ausi — APPROVED** [review]

**Comment by ausi** [comment]
This should be merged upstream for 5.3 and replace the current solution that exists there. 
Do you need an extra PR for that @leo ?

**Comment by leofeyer** [comment]
If the changes can be merged upstream, we don‘t need a separate PR. Only if the implementation differs between Contao 4 and Contao 5.

**Comment by leofeyer** [comment]
Thank you @fritzmg.

**Pull request merged**